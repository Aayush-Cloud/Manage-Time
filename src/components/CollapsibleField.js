// @flow
import * as React from 'react';
import {
  Animated,
  StyleSheet,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
/* eslint-disable import/no-unresolved, import/extensions */
import Textfield from './Textfield';
/* eslint-enable import/no-unresolved, import/extensions */

type Props = {
  children?: React.Node,
  headerText: string,
  meta: Object,
  onCollapse: () => void,
  onExpand: () => void,
}

type State = {
  collapsibleHeight: Animated.Value,
}

export interface Collapsible {
  collapse: () => void;
  expand: () => void;
}

/**
 * Renders a field that collapses/expands on click (like in iOS calendar app datepickers).
 *
 * Calls props.onCollapse/props.onExpand when collapsing/expanding by click but not
 * when collapsed/expanded programmatically by calling collapse()/exapand().
 *
 * Uses a <TextField /> to render the field itself (only thing that is visible when collapsed).
 * Additional props are passed to TextField.
 */
export default class CollapsibleField extends React.Component<Props, State>
  implements Collapsible {
  static defaultProps = {
    children: undefined,
    onCollapse: () => {},
    onExpand: () => {},
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      collapsibleHeight: new Animated.Value(0),
    };
    this._collapsed = true;
    this._childrenHeight = 0;
  }

  collapse = () => {
    this._animateCollapsibleHeight(0);
    this._collapsed = true;
  };

  expand = () => {
    this._animateCollapsibleHeight(this._childrenHeight);
    this._collapsed = false;
  };

  _animateCollapsibleHeight = (toValue: number) => {
    Animated.timing(this.state.collapsibleHeight, {
      toValue,
      duration: 220,
    }).start();
  };

  _toggle = () => {
    const { onCollapse, onExpand } = this.props;
    if (this._collapsed) {
      this.expand();
      onExpand();
    } else {
      this.collapse();
      onCollapse();
    }
  };

  _setChildrenHeight = (event: Object) => {
    this._childrenHeight = event.nativeEvent.layout.height;
  };

  _collapsed: boolean;
  _childrenHeight: number;

  render() {
    const { children, headerText, meta, ...rest } = this.props;
    // uses Textfield to have same look as other fields in forms
    return (
      <View style={styles.container}>
        <TouchableWithoutFeedback onPress={this._toggle}>
          <View>
            <Textfield
              editable={false}
              pointerEvents="none"
              meta={meta}
              value={headerText}
              {...rest}
            />
          </View>
        </TouchableWithoutFeedback>
        <Animated.View
          style={[styles.container, { height: this.state.collapsibleHeight }]}
        >
          <View
            onLayout={this._setChildrenHeight}
            style={styles.childrenContainer}
          >
            {children}
          </View>
        </Animated.View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    overflow: 'hidden',
  },
  childrenContainer: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
  },
});
