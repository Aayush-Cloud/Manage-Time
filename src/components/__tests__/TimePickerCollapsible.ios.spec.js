// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import CollapsibleField from '../CollapsibleField';
import TimePickerCollapsible from '../TimePickerCollapsible.ios';

const props = {
  starttime: {
    input: {
      onChange: jest.fn(),
      value: new Date(1),
    },
    meta: {
      submitFailed: true,
      error: 'starttime error',
    },
  },
  endtime: {
    input: {
      onChange: jest.fn(),
      value: new Date(2),
    },
    meta: {
      submitFailed: true,
      error: 'endtime error',
    },
  },
  placeholder: 'Some time',
};

describe('TimePickerCollapsible', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<TimePickerCollapsible {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls starttime.input.onChange when the value of the left picker changes', () => {
    const wrapper = shallow(<TimePickerCollapsible {...props} />);
    wrapper.find(CollapsibleField)
      .shallow()
      .find({ onDateChange: props.starttime.input.onChange })
      .simulate('dateChange');
    const { onChange } = props.starttime.input;
    expect(onChange.mock.calls.length).toBe(1);
  });

  it('calls endtime.input.onChange when the value of the right picker changes', () => {
    const wrapper = shallow(<TimePickerCollapsible {...props} />);
    wrapper.find(CollapsibleField)
      .shallow()
      .find({ onDateChange: props.endtime.input.onChange })
      .simulate('dateChange');
    const { onChange } = props.endtime.input;
    expect(onChange.mock.calls.length).toBe(1);
  });
});
