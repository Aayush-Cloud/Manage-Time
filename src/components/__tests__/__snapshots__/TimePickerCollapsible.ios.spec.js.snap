// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePickerCollapsible renders correctly 1`] = `
<View
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={undefined}
    testID={undefined}
  >
    <View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <TextInput
            clearButtonMode="while-editing"
            editable={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onCollapse={[Function]}
            onExpand={[Function]}
            onFocus={[Function]}
            placeholder="Some time"
            placeholderTextColor="rgba(0, 0, 0, 0.38)"
            style={
              Object {
                "alignItems": "flex-end",
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "Helvetica Neue",
                "fontSize": 16,
                "fontWeight": "400",
                "height": 28,
                "lineHeight": 16,
              }
            }
            value="1:00 AM - 1:00 AM"
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "height": 1,
            },
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.12)",
            },
          ]
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ff1744",
            "fontFamily": "Helvetica Neue",
            "fontSize": 14,
            "marginTop": 8,
          }
        }
      >
        endtime error
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "height": 0,
        "overflow": "hidden",
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <RCTDatePicker
            date={1}
            maximumDate={undefined}
            minimumDate={undefined}
            minuteInterval={5}
            mode="time"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "height": 216,
              }
            }
            timeZoneOffsetInMinutes={undefined}
          />
        </View>
        <View
          style={
            Object {
              "backgroundColor": "black",
              "height": 2,
              "width": 10,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <RCTDatePicker
            date={2}
            maximumDate={undefined}
            minimumDate={undefined}
            minuteInterval={5}
            mode="time"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "height": 216,
              }
            }
            timeZoneOffsetInMinutes={undefined}
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
